File: Algolympics/2021/i/i.cpp
#include <iostream>
#include <array>
#include <string>

int largest(int arr[], int n) 
{ 
    int i; 
    int max = arr[0]; 
    for (i = 1; i < n; i++) 
        if (arr[i] > max) 
            max = arr[i]; 
  
    return max; 
} 

int main()
{
    int t;
	scanf("%d", &t);
	while(t--){
		int n;
		int unsureCount = 0;
		std::string output = "";
		std::string addZero = "0 ";
		bool canBeZero = false;
		
		scanf("%d", &n);
		int* votes = (int*)calloc(n+1, sizeof(int));
		
		// getting votes
		for (int i = 0; i < n; i++){
			int vote;
			scanf("%d", &vote);
			if (vote == -1)
				unsureCount += 1;
			else
				votes[vote] += 1;
		}
		
		int mostSureCount = largest(votes, n+1);
		
		// check if -1 max
		if (unsureCount > mostSureCount){
			for (int i = 0; i <= n; i++){
				printf("%d ", i);
			}
		}
		else{ // mostSureCount is equal or higher than unsureCount
			// getting answer
			for (int i = 0; i <= n; i++){
				if (votes[i] + unsureCount > mostSureCount){
					if (i == 0) // 1 time check (just for 0th index), focus on else
						canBeZero = true;
					else
						output += std::to_string(i) + " ";
				}
				else if (votes[i] + unsureCount == mostSureCount)
					canBeZero = true;
			}
			if (canBeZero)
				output = addZero + output;
			 printf("%s ", output.c_str());
		}
		printf("\n");
	}
    return 0;
}

File: Algolympics/2021/i/i.py
t = int(input())
while(t):
    n = int(input())
    unsure_count = 0
    can_be_zero = False
    votes = []
    output = []
    
    # init votes list
    for i in range(n + 1):
        votes.append(0)
        
    # getting votes
    input_votes = list(map(int, input().split()))
    for vote in input_votes:
        if vote == -1:
            unsure_count += 1
        else:
            votes[vote] += 1
    
    max_vote = max(votes)
    
    # check if -1 max
    if unsure_count > max_vote:
        for i in range(n + 1):
            output.append(str(i))
        print(' '.join(output))
    # check if -1 == 0
    elif unsure_count == 0:
        count_maxes = 0
        crew_index = 0
        more_than_1 = False
        for i in range(n + 1):
            if votes[i] == max_vote:
                count_maxes += 1
                crew_index = i
            if count_maxes > 1:
                more_than_1 = True
                print('0')
                break
        if not more_than_1:
            print(crew_index)
    # -1 votes != 0 and -1 votes <= crew_vote
    else:
        for i in range(n + 1):
            if (votes[i] + unsure_count > max_vote):
                if i == 0:
                    can_be_zero = True
                else:
                    output.append(str(i))
            elif votes[i] + unsure_count == max_vote:
                can_be_zero = True
        if can_be_zero:
            output = ['0'] + output
        print(' '.join(output))
    
    t -= 1
